### PR TITLE
Feature/env variables override config 

### PR DESCRIPTION
**Pull request by pquerner**
### Description (*)

This PR attempts to implement Magento 2's feature of overwriting config values by the $_ENV variable.
The Magento 2 feature is documented here [1].

### Related Pull Requests
-

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#643

### Manual testing scenarios (*)

All steps assume you are using ddev as development enviroment.
1. In .ddev/config.yml add or fix `web_enviroment` setting: `web_environment: ["OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME=default", "OPENMAGE_CONFIG__WEBSITES__BASE__GENERAL__STORE_INFORMATION__NAME=website", "OPENMAGE_CONFIG__STORES__GERMAN__GENERAL__STORE_INFORMATION__NAME=store_german"]`
2. Go to adminhtml and check system configuration under "Configuration" > "General" > "Store-Information"+
3. Values from the database or default (config.xml) scope should be overwritten with given values from `web_environment`

(Else you can also set the variables in your server configuration, either in nginx or apache2 config)

### Automated testing scenarios (*)
I've given a PhpUnit test file to execute. The tests check if all scopes are correctly overwritten after calling the new EnvironmentLoader helper.

### Questions or comments

This feature was requested in #643 and @Flyingmana asked me to take a look at it.
Its use-case must be documented. @Flyingmana might be able to help out with that.

### ToDo / To be discussed further
I think the feature must be better implemented and documented (perhaps) than the original M2 feature-set.
For example its a little unclear (I didnt take a peek at M2 implementation) what should happen if you wish to "unset"/"delete" (?) a config value. 
For now I want to gather some needed feedback. 

Questions for example I have:

1. Is the implementation okay? (I somehow feel like Helper is the wrong class construct to use here, but I don't see a "Model" here either)
2. See above (override config should delete/unset?)

[1] https://experienceleague.adobe.com/docs/commerce-operations/configuration-guide/paths/override-config-settings.html

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)
 - [ ] Add yourself to contributors list
 
(did none here, if someone can fix it pre-merge that would be great. Last time I executed that `yarn all-contributors` I didnt get added properly)

### CODE REVIEW

**Comment by fballiano** [comment]
Unit tests fail on php8 https://github.com/OpenMage/magento-lts/actions/runs/7957571133/job/21720971313?pr=3842

**Comment by pquerner** [comment]
I dont have these fails locally on 8.1.27. Not sure whats broken.

File: dev/tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+     */
+    public function testEnvOverrides(array $config)
+    {
+        error_reporting(0);
+        $xmlStruct = $this->getTestXml();
+
+        $xmlDefault = new Mage_Core_Model_Config($xmlStruct);
+        $xmlDefault = $xmlDefault->loadModulesConfiguration('config.xml', $xmlDefault);
+        $xml = new Mage_Core_Model_Config($xmlStruct);
+        $xml = $xml->loadModulesConfiguration('config.xml', $xml);
```
Comment (kiatng) [inline dev/tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:50]: To fix the error 

> TypeError: Cannot access offset of type string on string

I tested the following code snippet (**I do not really know what I am doing, just trial and error**):

```suggestion
        $xml = new Varien_Simplexml_Config();
        $xml->loadString($xmlStruct);
        $config = Mage::getConfig()->loadModulesConfiguration('config.xml', $xml);
        $r = $config->getNode('default/general/store_information/name');
```

$r is
```
 object(Varien_Simplexml_Element)#177 (1) {
  [0] => string(12) "test_default"
}
```

Hope it helps you to fix the errors.

**Comment by pquerner** [comment]
Unittests have been fixed.

**Comment by fballiano** [comment]
It seems something went wrong with the rebase because the PR has 26 modified files :-\

**Comment by pquerner** [comment]
Oh darn, you're right. Let me see if I can fix that.

**Comment by pquerner** [comment]
No idea, and I don't have a backup branch sadly. Maybe close and open again from the branch? The remote branch on my fork doesnt have that many changed files. Or do you see another way @fballiano ?

**Comment by pquerner** [comment]
Re-opened the PR here https://github.com/OpenMage/magento-lts/pull/3863

**Pull request closed without merging**